presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.facebook.presto.sql.tree.WithQuery;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.DUPLICATE_RELATION;
+import static com.facebook.presto.sql.analyzer.SemanticExceptions.createAmbiguousAttributeException;
+import static com.facebook.presto.sql.analyzer.SemanticExceptions.createMissingAttributeException;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Sets.newIdentityHashSet;
+import static java.util.Objects.requireNonNull;
+
+public class Scope
```

Comment:
I wanted to make immutable, but it was not convenient for all the use cases. Now with `Builder` is it is better.

Now, once you have asked, I made it immutable. To achieve that I had to move the field `columnReferences` to `Analysis` (anyway it is a better place for that). 

Take a look `StatementAnalyzer#visitQuery` and `StatementAnalyzer#analyzeWith` methods. These are the only places where immutability of `Scopes` makes troubles, because I need to create partial `Scope` objects to analyze nested nodes. To me it is acceptable, but I would like to hear your voice about it.